http-stream2: Add request body length option

lenticularis39 opened the pull request:
This enables sending the body continuously as it is received from the
transport instead of storing it in memory and sending it with the header
at once.

---

This is a simple patch primarily aiming to provide file upload support via HTTP post requests, for example for use with REST API services like Podman.

mvollmer commented:
Thanks, this looks good!

Let's add some tests to test-httpstream.c.

mvollmer reviewed (changes requested):
This needs a test in test-httpstream.c

lenticularis39 commented:
After looking into this it looks like the tests would require either modifying `CockpitWebServer` or writing a custom server, since `CockpitWebServer` doesn't support POST requests, therefore `body-length` can't be tested using it. I'm not sure if this is the right path to go considering `CockpitWebServer` is not used only in tests, so a copy would likely be necessary for the test.

martinpitt commented:
FTR, if an integration test is any easier, that would be fine I think. It is much closer to how a "real" page would use that API, and thus would serve as a code example. You can either make up something dynamic with `b.eval_js()`, or modify one of the playground pages to do what you want.

lenticularis39 commented:
Added tests (using a modified version of the Cockpit web server that can accept POST requests) and it seems to work, the file structure is possibly not the best though (maybe keepin the test web server in src/bridge instead of a separate test directory would be better).

martinpitt commented:
[unit tests failed](https://github.com/cockpit-project/cockpit/pull/14849/checks?check_run_id=2165111031) with `test-httpstream process failed: 124`, which means "timeout".

martinpitt commented:
@lenticularis39 : Thanks for the detailed test! (I didn't review it in detail yet).

Can you please rebase your branch against current master? That may or may not help with the unit tests, but certainly it makes it easier to test.

I started the fedora-33 integration test to give it some beating.

lenticularis39 commented:
> [unit tests failed](https://github.com/cockpit-project/cockpit/pull/14849/checks?check_run_id=2165111031) with `test-httpstream process failed: 124`, which means "timeout".

Ah, I see: the test sends about 300 MB of data though the stream and then checks it, which takes about 150 seconds when running natively; under Valgrind it's much slower of course.

martinpitt commented:
@lenticularis39 : You can either call
```c
  if (cockpit_test_skip_slow ())
    return;
```
in that test (some tests already do that), or have it include `cockpithacks.h` and reduce the data size `if (RUNNING_ON_VALGRIND)`. The latter is of course better, as it will then still check for memory leaks in the code.

lenticularis39 commented:
> @lenticularis39 : You can either call
> 
> ```c
>   if (cockpit_test_skip_slow ())
>     return;
> ```
> 
> in that test (some tests already do that), or have it include `cockpithacks.h` and reduce the data size `if (RUNNING_ON_VALGRIND)`. The latter is of course better, as it will then still check for memory leaks in the code.

It's good to know `RUNNING_ON_VALGRIND` is available. I did the size reduction by checking for Valgrind in LD_PRELOAD, because I wasn't sure if building without valgrind.h is supported and didn't want to break compilation without it; knowing it's used at other places I'll update the checking code.

martinpitt commented:
@lenticularis39 : Right, cockpithacks.h has a fallback without valgrind, the macro is just `(0)` then.

lenticularis39 commented:
Rebased and added the valgrind check.

Note: If valgrind is installed, but valgrind-devel is not, `RUNNING_ON_VALGRIND` falls back to `0`, which might be an unpleasant surprise for someone doing `make check-memory` without valgrind-devel and not knowing about it. Maybe this could be somehow mentioned in the documentation.

martinpitt commented:
@allisonkarlitskaya : Gentle review ping?

allisonkarlitskaya commented on src/bridge/cockpithttpstream.c:
```diff
@@ -215,6 +215,11 @@ typedef struct _CockpitHttpStream {
+
+  /* Stores number of bytes in the request waiting to be sent */
+  /* Only valid when body_length is set (otherwise the request is sent at once) */
+  gulong request_buffer_size;
```
I don't understand why we do this rather than just sending immediately, once the header is sent.  `cockpit_stream_write()` is also doing its own buffering.

Also: the buffering in the stream seems to already be throttling-enabled.  It might be nice to write a test to make sure this is functioning properly for large POST requests with e.g. a fake webserver which lets the connection fill up, and making sure that the connection exerts back-pressure properly.  Look at cockpitflow for more information about how that should work.

lenticularis39 replied:
Good point. Simply sending the received data via `cockpit_stream_write` on receive seems to be very slow when sending small pieces of data multiple times (like the "HelloWorld" string in the tests), though. I'll have a look at what exactly is happening there.

allisonkarlitskaya commented on src/bridge/cockpithttpstream.c:
```diff
@@ -726,11 +750,9 @@ send_http_request (CockpitHttpStream *self)
   const gchar *value;
-  GList *request = NULL;
-  GList *names = NULL;
+  g_autoptr(GList) names = NULL;
```
is this even being used anywhere?

lenticularis39 replied:
It's used for storing the list of `headers` option's members.

allisonkarlitskaya reviewed (changes requested):
Thanks very much for this patch.  Sorry for taking so long to review this!

In general the idea looks good, but I have one nitpick about the extra layer of buffering.  I'd be happy to take another look after that is dropped (or happy also to hear why it shouldn't be dropped).

About the name of the option: when i think about the "body-length", i'd normally imagine that i think about the reply.  maybe "request-body-length", or even "post-body-length" and restrict it to only working with POST?

martinpitt commented:
This is quite stale, has several open review comments, and now also got overtaken by our rewriting of the bridge in Python. So this doesn't apply at all any more. On the plus side, writing unit tests with pytest and the full Python API available should now be a lot simpler, too.

Are you still interested in this @lenticularis39 or shall we close this? (Sorry..)

lenticularis39 commented:
I was thinking about this PR for a while, so I didn't close it with the others, but I probably won't have the motivation to work on it. Also, I don't like the implementation that much, so I'd probably do it again anyway.
